allennlp/data/semparse/type_declarations/type_declaration.py

```diff
@@ -1,28 +1,111 @@
 """
-This module defines some classes that are generally useful for defining a type system for a new domain. We
-inherit the type logic in ``nltk.sem.logic`` and add some functionality on top of it here. There are two main
-improvements:
+This module defines some classes that are generally useful for defining a type system for a new
+domain. We inherit the type logic in ``nltk.sem.logic`` and add some functionality on top of it
+here. There are two main improvements:
 1) Firstly, we allow defining multiple basic types with their own names (see ``NamedBasicType``).
-2) Secondly, we allow defining function types that have placeholders in them (see ``PlaceholderType``).
-We also extend NLTK's ``LogicParser`` to define a ``DynamicTypeLogicParser`` that knows how to deal with the
-two improvements above.
+2) Secondly, we allow defining function types that have placeholders in them (see
+``PlaceholderType``).
+We also extend NLTK's ``LogicParser`` to define a ``DynamicTypeLogicParser`` that knows how to deal
+with the two improvements above.
 """
 from typing import Dict, List, Optional, Set, Tuple, Union
 from collections import defaultdict
-import re
 
 from overrides import overrides
 from nltk.sem.logic import Expression, ApplicationExpression, ConstantExpression, LogicParser, Variable
-from nltk.sem.logic import Type, BasicType, ComplexType, ANY_TYPE
+from nltk.sem.logic import Type, BasicType, ComplexType as NltkComplexType, ANY_TYPE
 
 START_SYMBOL = '@START@'
 
 
+class ComplexType(NltkComplexType):
+    """
+    In NLTK, a ``ComplexType`` is a function.  These functions are curried, so if you need multiple
+    arguments for your function you nest ``ComplexTypes``.  That currying makes things difficult
+    for us, and we mitigate the problems by adding ``return_type`` and ``argument_type`` functions
+    to ``ComplexType``.
+    """
+    def return_type(self):
+        """
+        Gives the final return type for this function.  If the function takes a single argument,
+        this is just ``self.second``.  If the function takes multiple arguments and returns a basic
+        type, this should be the final ``.second`` after following all complex types.  That is the
+        implementation here in the base class.  If you have a higher-order function that returns a
+        function itself, you need to override this method.
+        """
+        return_type = self.second
+        while isinstance(return_type, ComplexType):
+            return_type = return_type.second
+        return return_type
+
+    def argument_types(self):
+        """
+        Gives the types of all arguments to this function.  For functions returning a basic type,
+        we grab all ``.first`` types until ``.second`` is no longer a ``ComplexType``.  That logic
+        is implemented here in the base class.  If you have a higher-order function that returns a
+        function itself, you need to override this method.
```

Comment:
It doesn't, actually, because `self.first` or `remaining_type.first` will return a function type, and everything is fine.  The `argmax` and `argmin` functions take functions as arguments, and this logic works fine.